File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/block/SurfBukkitNmsBlockPacketsImpl.java
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.block;

import static com.google.common.base.Preconditions.checkNotNull;

import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.PacketOperation;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.block.SurfBukkitNmsBlockPackets;
import dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.PacketOperationImpl;
import dev.slne.surf.surfapi.bukkit.server.nms.NmsUtil;
import io.papermc.paper.math.BlockPosition;
import javax.annotation.ParametersAreNonnullByDefault;
import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
import org.bukkit.block.data.BlockData;

@SuppressWarnings("UnstableApiUsage")
@ParametersAreNonnullByDefault
public final class SurfBukkitNmsBlockPacketsImpl implements SurfBukkitNmsBlockPackets, NmsUtil {

  @Override
  public PacketOperation updateBlockData(BlockPosition position, BlockData blockData) {
    checkNotNull(position, "position");
    checkNotNull(blockData, "blockData");
    return PacketOperationImpl.simple(
        player -> new ClientboundBlockUpdatePacket(toNms(position), toNms(blockData)));
  }

  @Override
  public PacketOperation resetBlock(BlockPosition position) {
    checkNotNull(position, "position");

    return PacketOperationImpl.simple(
        player -> new ClientboundBlockUpdatePacket(toNms(player).level(), toNms(position)));
  }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/block/SurfBukkitNmsBlockPacketsImpl.kt
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.block

import com.google.auto.service.AutoService
import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.block.SurfBukkitNmsBlockPackets
import dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.PacketOperationImpl
import dev.slne.surf.surfapi.bukkit.server.nms.toNms
import dev.slne.surf.surfapi.core.api.util.checkInstantiationByServiceLoader
import io.papermc.paper.math.BlockPosition
import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket
import org.bukkit.block.data.BlockData

@AutoService(SurfBukkitNmsBlockPackets::class)
@NmsUseWithCaution
class SurfBukkitNmsBlockPacketsImpl : SurfBukkitNmsBlockPackets {
    init {
        checkInstantiationByServiceLoader()
    }

    override fun updateBlockData(position: BlockPosition, blockData: BlockData) =
        PacketOperationImpl.simple {
            ClientboundBlockUpdatePacket(
                position.toNms(),
                blockData.toNms()
            )
        }


    override fun resetBlock(position: BlockPosition) = PacketOperationImpl.simple {
        ClientboundBlockUpdatePacket(
            it.toNms().level(),
            position.toNms()
        )
    }
}

